nixos/modules/services/networking/frr.nix

```diff
@@ -0,0 +1,201 @@
+        example =
+          let
+            examples = {
+              rip = ''
+                router rip
+                  network 10.0.0.0/8
+              '';
+
+              ospf = ''
+                router ospf
+                  network 10.0.0.0/8 area 0
+              '';
+
+              bgp = ''
+                router bgp 65001
+                  neighbor 10.0.0.1 remote-as 65001
+              '';
+            };
+          in
+            examples.${service} or "";
+        description = ''
+          ${daemonName service} configuration statements.
+        '';
+      };
+
+      vtyListenAddress = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Address to bind to for the VTY interface.
+        '';
+      };
+
+      vtyListenPort = mkOption {
+        type = types.nullOr types.int;
+        default = null;
+        description = ''
+          TCP Port to bind to for the VTY interface.
+        '';
+      };
+    };
+
+in
+
+{
+
+  ###### interface
+  imports = [
+    {
+      options.services.frr = {
+        zebra = (serviceOptions "zebra") // {
+          enable = mkOption {
+            type = types.bool;
+            default = any isEnabled services;
+            description = ''
+              Whether to enable the Zebra routing manager.
+
+              The Zebra routing manager is automatically enabled
+              if any routing protocols are configured.
+            '';
+          };
+        };
+      };
+    }
+    { options.services.frr = (genAttrs services serviceOptions); }
+  ];
+
+  ###### implementation
+
+  config = mkIf (any isEnabled allServices) {
+
+    environment.systemPackages = [
+      pkgs.frr # for the vtysh tool
+    ];
+
+    users.users.frr = {
+      description = "FRR daemon user";
+      isSystemUser = true;
+      group = "frr";
+    };
+
+    users.groups = {
+      frr = {};
+      # Members of the frrvty group can use vtysh to inspect the FRR daemons
+      frrvty = { members = [ "frr" ]; };
+    };
+
+    systemd.services =
+      let
+        frrService = service:
+          let
+            scfg = cfg.${service};
+            daemon = daemonName service;
+          in
+            nameValuePair daemon ({
+              wantedBy = [ "multi-user.target" ];
+              after = [ "network-pre.target" "systemd-sysctl.service" ];
+              wants = [ "network.target" ];
+
+              serviceConfig = {
+                Type = "forking";
+                PIDFile = "/run/frr/${daemon}.pid";
+                ExecStart = "@${pkgs.frr}/libexec/frr/${daemon} ${daemon} -d -f ${configFile service}"
+                  + optionalString (scfg.vtyListenAddress != "") " -A ${scfg.vtyListenAddress}"
+                  + optionalString (scfg.vtyListenPort != null) " -P ${toString scfg.vtyListenPort}";
+                ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+                Restart = "on-abnormal";
```

Comment:
I'm not convinced this systemd structure is the right approach here.

- There's no proper supervision of unresponsive daemons. 
- Starting daemons must happen in a particular order, specifically zebra must be started first
- Restarting requires careful separation of daemons into phases.

Upstream frr comes with a systemd unit that simply runs watchfrr, handling all of those above points.

Reply:
Further, this systemd unit *must* reloadIfChanged, otherwise this tears down all routing protocol sessions, removes routes from RIB, etc...

Reply:
You're right, watchfrr should be used and this needs cleanup and rework. I don't know when I can do it yet.

Reply:
> There's no proper supervision of unresponsive daemons.

What does frr do in that case? I don't grok how `watchfrr` recognizes unresponsive daemons.

> Starting daemons must happen in a particular order, specifically zebra must be started first

That's easy, we let everything run after zebra. But I'm not sure what to do about all the other daemons. The dependencies are a list, not a tree. https://github.com/FRRouting/frr/blob/5af4491693676f7c52aeed3a981e6621a3a8bba4/tools/frrcommon.sh.in#L34-L38

> Restarting requires careful separation of daemons into phases.

Restarting with NixOS will be tough, since on a rebuild switch nixos will call `systemctl restart zebra.service ospfd.service bgpd.service` and systemd will not care about ordering. No idea how to resolve this.